eng/Tools.props

```diff
@@ -59,14 +59,23 @@
     <RepoTool Include="dotnet-reportgenerator-globaltool" Version="$(DotNetReportGeneratorGlobalToolPackageVersion)" />
   </ItemGroup>
 
-  <!--
-    Temporarily opt-in as dotnet restore interactive doesn't seem to work.
-    https://github.com/NuGet/Home/issues/7812. Use nuget restore meanwhile.
-  -->
+  <!-- Opt-in target to restore internal tools with interactive authentication. -->
   <Target Name="InitOptionalTools"
           Condition="'$(RestoreInternalTools)' == 'true'"
           AfterTargets="Restore">
 
+    <PropertyGroup>
+      <CredentialsProviderScriptUrl Condition="'$(OS)' == 'Windows_NT'">https://raw.githubusercontent.com/Microsoft/artifacts-credprovider/master/helpers/installcredprovider.ps1</CredentialsProviderScriptUrl>
+      <CredentialsProviderScriptUrl Condition="'$(OS)' != 'Windows_NT'">https://raw.githubusercontent.com/Microsoft/artifacts-credprovider/master/helpers/installcredprovider.sh</CredentialsProviderScriptUrl>
+    </PropertyGroup>
+
+    <DownloadFile SourceUrl="$(CredentialsProviderScriptUrl)" DestinationFolder="$(ArtifactsDir)">
```

Comment:
I would not expect this to go to ArtifactsDir - I think it would be better in `/.tools/CredentialsProvider`

Reply:
Also, we shouldn't download it if we already have it.

Reply:
Where does the installer install the cred provider?

Reply:
> Where does the installer install the cred provider?

under %userprofile%\.nuget\plugins

> I would not expect this to go to ArtifactsDir - I think it would be better in /.tools/CredentialsProvider

What is .tools? :O We don't have that in corefx.

> Also, we shouldn't download it if we already have it.

Redownloading makes sure to update the provider.

Reply:
Does the installer script respect `NUGET_PACKAGES` variable? If not, can we pass the path in?

It's where Arcade downloads tools (e.g. vswhere). https://github.com/dotnet/arcade/blob/master/eng/common/tools.ps1#L511
So you do have it in CoreFX - perhaps not using it for anything right now.

> Redownloading makes sure to update the provider.

Does the script have no version that we can check?

Reply:
> Does the script have no version that we can check?

No that script is hosted directly on GH, it is not versioned. That's also the reason why I made this target an opt-in as it could break if someone moves the script.

> Does the installer script respect NUGET_PACKAGES variable? If not, can we pass the path in?

As you can see in the script, the plugin path is hardcoded: https://raw.githubusercontent.com/Microsoft/artifacts-credprovider/master/helpers/installcredprovider.sh. You could submit a PR that fixes that.

Reply:
If we moved this functionality to Arcade we would probably want to implement this directly in `tools.ps1|sh`, instead of downloading the script - seems like the script is just downloading a binary and placing it to the directory.

@nkolev92 @rrelyea Is there a better way to acquire credential provider for NuGet? Should `NuGet.exe`/`dotnet restore` do that automatically based on a configuration setting in `NuGet.config` (something like `<UseCredentialProvider>true</UseCredentialProvider>`).

Reply:
There's a separation between the nuget extensibility point for cross platform credential providers and the plugins themselves. 

In Visual Studio (msbuild) the ADO plugin is already installed alongside NuGet, but cross platform with dotnet.exe that's not the case. 

The decision was to not add services support in the dotnet.exe by default. 

Not sure there's a better way today. We are planning on allowing the plugins to be dotnet tools eventually, but that's not happening until dotnet.exe introduces local/repo tools.

Reply:
Maintaining separation is fine. However, I think I should be able to specify what plugins I want to use in `NuGet.config` and let NuGet "restore" them before it restores packages. So perhaps the configuration would look like:

```xml
<plugins>
  <add name="Microsoft.NuGet.CredentialProvider" version="x.y.z" feed="..." />
</plugins
```

Reply:
@ViktorHofer Feel free to file an Arcade issue on adding this feature to Arcade SDK referring to this conversation and link it from a comment in your CoreFX change.

Reply:
https://github.com/dotnet/arcade/issues/2416

Reply:
Feel free to file an issue about that @tmat :)

I agree the current experience is rather rudimentary, but we just didn't invest into designing an experience about something easy to script.

We have not asked credential provider authors to package their plugins yet, so I'd be hesitant to add an experience like the above. 
Another reason why we decided against that initially was the fact that you have multiple nuget configs, and different nuget.configs might be used for different projects. The plugins however are process/restore operation wide.

Personally I think, having them as repo tools is a better experience.

Reply:
Yeah, repo tools would probably work as well.

Reply:
> It's where Arcade downloads tools (e.g. vswhere). https://github.com/dotnet/arcade/blob/master/eng/common/tools.ps1#L511
So you do have it in CoreFX - perhaps not using it for anything right now.

Could we make that variable available in the arcade as an msbuild variable so that corefx can use it in their msbuild logic?

Reply:
> I would not expect this to go to ArtifactsDir - I think it would be better in /.tools/CredentialsProvider

done.

Reply:
> Does the installer script respect NUGET_PACKAGES variable? If not, can we pass the path in?

https://github.com/Microsoft/artifacts-credprovider/issues/89

> We would also probably want to specify the required version explicitly (instead of always using the latest) so that our builds are deterministic.

https://github.com/Microsoft/artifacts-credprovider/issues/88